x-pack/platform/plugins/shared/cloud/server/plugin.ts

```diff
@@ -200,6 +205,63 @@ export class CloudPlugin implements Plugin<CloudSetup, CloudStart> {
       projectType,
       orchestratorTarget,
     });
+    const basePath = core.http.basePath.serverBasePath;
+    core.http.resources.register(
+      {
+        path: '/app/cloud/onboarding',
+        validate: {
+          query: schema.maybe(
+            schema.object(
+              {
+                next: schema.maybe(schema.string()),
+                onboarding_token: schema.maybe(schema.string()),
+              },
+              { unknowns: 'ignore' }
+            )
+          ),
+        },
+        security: {
+          authz: {
+            enabled: false,
+            reason: 'any user can open kibana from cloud ui',
+          },
+        },
+      },
+      async (context, request, response) => {
+        const { uiSettings } = await context.core;
+        const defaultRoute = await uiSettings.client.get<string>('defaultRoute', { request });
+        const nextCandidateRoute = parseNextURL(request.url.href);
+
+        const route = nextCandidateRoute === '/' ? defaultRoute : nextCandidateRoute;
+        // need to get reed of ../../ to make sure we will not be out of space basePath
+        const normalizedRoute = new URL(route, 'https://localhost');
+
+        const queryOnboardingToken = request.url.searchParams.get('onboarding_token');
+        const solutionType = this.config.onboarding?.default_solution;
+        if (queryOnboardingToken) {
+          core
+            .getStartServices()
+            .then(async ([coreStart]) => {
+              const soClient = coreStart.savedObjects.getScopedClient(request, {
```

Comment:
The `savedObjects` should come from `const { uiSettings, savedObjects } = await context.core; //line 231` and that means you can likely get rid of `core.getStartServices().then` 
```suggestion
              const soClient = savedObjects.getClient({
```